python-backend/app/api/permits/permit_amendment/resources/permit_amendment_document.py

```diff
@@ -64,71 +62,76 @@ def post(self, permit_guid=None, document_guid=None):
-            data=data,
+            data=response_data,
             cookies=request.cookies,
         )
 
         response = Response(resp.content, resp.status_code, resp.raw.headers.items())
         return response
 
+    def _parse_upload_folders(self, mine_guid):
+        folder = f'mines/{str(mine_guid)}/permits'
+        pretty_folder = f'mines/{mine_guid}/permits'
+
+        return folder, pretty_folder
+
+    def _parse_request_metadata(self):
+        request_metadata = request.headers.get("Upload-Metadata")
+        metadata = {}
+        if not request_metadata:
+            return metadata
+
+        for key_value in request_metadata.split(","):
+            (key, value) = key_value.split(" ")
+            metadata[key] = base64.b64decode(value).decode("utf-8")
+
+        return metadata
+
+    @api.marshal_with(PERMIT_AMENDMENT_DOCUMENT_MODEL, code=201)
     @requires_role_mine_create
-    def put(self, permit_amendment_guid, document_guid=None, permit_guid=None):
+    def put(self, permit_amendment_guid, permit_amendment_document_guid=None, permit_guid=None):
         permit_amendment = PermitAmendment.find_by_permit_amendment_guid(permit_amendment_guid)
         if not permit_amendment:
-            return self.create_error_payload(404, 'Permit amendment not found'), 404
+            raise NotFound('Permit amendment not found.')
 
         data = self.parser.parse_args()
         if data.get('document_manager_guid'):
             # Register and associate a new file upload
             filename = data.get('filename')
             if not filename:
-                return self.create_error_payload(400,
-                                                 'Must supply filename for new file upload'), 400
+                raise BadRequest('Must supply filename for new file upload.')
 
             new_pa_doc = PermitAmendmentDocument(
                 mine_guid=permit_amendment.permit.mine_guid,
                 document_manager_guid=data.get('document_manager_guid'),
                 document_name=filename)
 
-            permit_amendment.documents.append(new_pa_doc)
+            permit_amendment.related_documents.append(new_pa_doc)
             permit_amendment.save()
         else:
-            return self.create_error_payload(
-                400, 'Must provide the doc manager guid for the newly uploaded file'), 400
+            raise BadRequest('Must provide the doc manager guid for the newly uploaded file.')
+
+        return new_pa_doc
 
-        return permit_amendment.json()
 
+class PermitAmendmentDocumentResource(Resource, UserMixin):
     @requires_role_mine_create
-    def delete(self, permit_amendment_guid, document_guid=None, permit_guid=None):
-        if not document_guid:
-            return self.create_error_payload(400, 'must provide document_guid to be unlinked'), 400
+    @api.response(204, 'Successfully deleted.')
+    def delete(self, permit_amendment_guid, permit_amendment_document_guid=None, permit_guid=None):
```

Comment:
should these be optional variables? looks like they should be required.

Reply:
Agree, changed to required here and in a few other places. Does flaskrestplus enforce these if the default is not None? Do we need to keep the conditional validation below?

Reply:
if they are not defaulted, then all of the URL parameter defined i the path need to exist. correct no more None check.

Reply:
Updated and removed None check